File: leetcode/150-evaluate-reverse-polish-notation/main.go
package main

import (
	"fmt"
	"strconv"
)

func evalRPN(tokens []string) int {
	if len(tokens) == 0 {
		return 0
	}
	var stack []int
	for i := 0; i < len(tokens); i++ {
		curr := tokens[i]
		if curr != "+" && curr != "-" && curr != "*" && curr != "/" {
			v, _ := strconv.Atoi(curr)
			stack = append(stack, v)
		} else {
			a := stack[len(stack)-2]
			b := stack[len(stack)-1]
			stack = stack[:len(stack)-2]
			if curr == "+" {
				temp := a + b
				stack = append(stack, temp)
			} else if curr == "-" {
				temp := a - b
				stack = append(stack, temp)
			} else if curr == "*" {
				temp := a * b
				stack = append(stack, temp)
			} else if curr == "/" {
				temp := a / b
				stack = append(stack, temp)
			}
		}
	}
	return stack[0]
}

func main() {
	fmt.Println(evalRPN([]string{"2", "1", "+", "3", "*"}))
	fmt.Println(evalRPN([]string{"10", "6", "9", "3", "+", "-11", "*", "/", "*", "17", "+", "5", "+"}))
}

File: leetcode/150-evaluate-reverse-polish-notation/main.py
"""
    1st: stack

    Time    O(N)
    Space   O(N)
    48 ms, faster than 98.82%
"""
class Solution(object):
    def evalRPN(self, tokens):
        stack = []
        for t in tokens:
            if t == '+' or t == '-' or t == '*' or t == '/':
                b = stack.pop()
                a = stack.pop()
                if t == '+':
                    stack.append(a + b)
                elif t == '-':
                    stack.append(a - b)
                elif t == '*':
                    stack.append(a * b)
                elif t == '/':
                    if (a < 0 and b < 0) or (a >= 0 and b > 0):
                        stack.append(a // b)
                    elif a < 0:
                        stack.append(-(abs(a) // b))
                    elif b < 0:
                        stack.append(-(a // abs(b)))
            else:
                stack.append(int(t))
        return stack[0]